core/src/main/java/io/questdb/cairo/vm/MemoryCARWSpillable.java

```diff
@@ -0,0 +1,283 @@
+        return mem.getPageAddress(pageIndex);
+    }
+
+    @Override
+    public int getPageCount() {
+        return mem.getPageCount();
+    }
+
+    @Override
+    public long getPageSize() {
+        return getExtendSegmentSize();
+    }
+
+    @Override
+    public CharSequence getStr(long offset) {
+        return mem.getStr(offset);
+    }
+
+    @Override
+    public CharSequence getStr2(long offset) {
+        return mem.getStr2(offset);
+    }
+
+    /**
+     * Updates append pointer with address for the given offset. All put* functions will be
+     * appending from this offset onwards effectively overwriting data. Size of virtual memory remains
+     * unaffected until the moment memory has to be extended.
+     *
+     * @param offset position from 0 in virtual memory.
+     */
+    public void jumpTo(long offset) {
+        checkAndExtend(mem.getAddress() + offset);
+        mem.jumpTo(offset);
+    }
+
+    @Override
+    public long offsetInPage(long offset) {
+        return mem.offsetInPage(offset);
+    }
+
+    @Override
+    public int pageIndex(long offset) {
+        return mem.pageIndex(offset);
+    }
+
+    @Override
+    public void putLong256(@NotNull CharSequence hexString, int start, int end) {
+        mem.putLong256(hexString, start, end);
+    }
+
+    @Override
+    public long resize(long size) {
+        extend(size);
+        return mem.getAddress();
+    }
+
+    @Override
+    public void shiftAddressRight(long shiftRightOffset) {
+        mem.shiftAddressRight(shiftRightOffset);
+    }
+
+    @Override
+    public long size() {
+        return mem.size();
+    }
+
+    /**
+     * Skips given number of bytes. Same as logically appending 0-bytes. Advantage of this method is that
+     * no memory write takes place.
+     *
+     * @param bytes number of bytes to skip
+     */
+    @Override
+    public void skip(long bytes) {
+        checkAndExtend(getAppendAddress() + bytes);
+        mem.skip(bytes);
+    }
+
+    /**
+     * Reset the appendOffset and resize to 1 page
+     */
+    @Override
+    public void truncate() {
+        if(isSpilled){
+            mem.close();
+            ff.remove(spillPath);
+            mem = initCARW();
+            isSpilled = false;
+        }
+        mem.truncate();
+        assert mem.size() == pageSize;
+    }
+
+    @Override
+    public void zero() {
+        mem.zero();
+    }
+
+    private void checkAndExtend(long address) {
+        if (address <= mem.getAddress() + mem.size()) {
+            return;
+        }
+        long sz = address - mem.getAddress();
+        if (!isSpilled && (sz > diskSpillThreshold)){
+
+            MemoryCARW cmarw = initCMARW(sz, mem.getAppendOffset());
+            synchronized (this){
+                Vect.memmove(cmarw.getAddress(), mem.getAddress(), mem.size());
+                LOG.debug().$("Start spilling. curSz: ").$(mem.size()).$(", requestedSz: ").$(sz).$(", newSz: ").$(cmarw.size());
```

Comment:
A trailing `.$()`/`.I$()` call must end any logging call:
```suggestion
                LOG.debug().$("Start spilling. curSz: ").$(mem.size()).$(", requestedSz: ").$(sz).$(", newSz: ").$(cmarw.size()).$();
```